manager/state/raft/storage/storage.go

```diff
@@ -0,0 +1,382 @@
+
+	if !fileutil.Exist(snapDir) {
+		// If snapshots created by the etcd-v2 code exist, read the latest snapshot
+		// and write it encoded to the new path.  The new path prevents etc-v2 creating
+		// snapshots that are visible to us, but not encoded and out of sync with our
+		// WALs, after a downgrade.
+		legacySnapDir := e.legacySnapDir()
+		if fileutil.Exist(legacySnapDir) {
+			if err := MigrateSnapshot(legacySnapDir, snapDir, OriginalSnap, snapFactory); err != nil {
+				return nil, err
+			}
+		} else if err := os.MkdirAll(snapDir, 0700); err != nil {
+			return nil, errors.Wrap(err, "failed to create snapshot directory")
+		}
+	}
+
+	var (
+		snapshotter Snapshotter
+		walObj      WAL
+		err         error
+		result      = &RaftDataFromDisk{}
+	)
+
+	// Create a snapshotter and load snapshot data
+	snapshotter = snapFactory.New(snapDir)
+	result.Snapshot, err = snapshotter.Load()
+	if err != nil && err != snap.ErrNoSnapshot {
+		return nil, err
+	}
+
+	walFactory := NewWALFactory(encoder, decoder)
+	var walsnap walpb.Snapshot
+	if result.Snapshot != nil {
+		walsnap.Index = result.Snapshot.Metadata.Index
+		walsnap.Term = result.Snapshot.Metadata.Term
+	}
+
+	if !wal.Exist(walDir) {
+		// If wals created by the etcd-v2 wal code exist, read the latest ones based
+		// on this snapshot and encode them to wals in the new path to avoid adding
+		// backwards-incompatible entries to those files.
+		legacyWALDir := e.legacyWALDir()
+		if !wal.Exist(legacyWALDir) {
+			return nil, ErrNoWAL
+		}
+
+		if err = MigrateWALs(legacyWALDir, walDir, OriginalWAL, walFactory, walsnap, log.G(ctx)); err != nil {
+			return nil, err
+		}
+	}
+
+	walObj, result.Metadata, result.HardState, result.Entries, err = ReadRepairWAL(walDir, walsnap, walFactory, log.G(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	e.snapshotter = snapshotter
+	e.wal = walObj
+
+	return result, nil
+}
+
+// BootstrapNew creates a new snapshotter and WAL writer, expecting that there is nothing on disk
+func (e *EncryptedRaftLogger) BootstrapNew(raftNode *api.RaftMember) ([]byte, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	encoder, decoder := encryption.GetDefaults(e.EncryptionKey)
+	walFactory := NewWALFactory(encoder, decoder)
+
+	for _, dirpath := range []string{e.walDir(), e.snapDir()} {
+		if err := os.MkdirAll(dirpath, 0700); err != nil {
+			return nil, errors.Wrap(err, "failed to create WAL directory")
+		}
+	}
+
+	metadata, err := raftNode.Marshal()
+	if err != nil {
+		return nil, errors.Wrap(err, "error marshalling raft node")
+	}
+	e.wal, err = walFactory.Create(e.walDir(), metadata)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create WAL")
+	}
+
+	e.snapshotter = NewSnapFactory(encoder, decoder).New(e.snapDir())
+	return metadata, nil
+}
+
+func (e *EncryptedRaftLogger) legacyWALDir() string {
+	return filepath.Join(e.StateDir, "wal")
+}
+
+func (e *EncryptedRaftLogger) walDir() string {
+	return filepath.Join(e.StateDir, "wal-v3")
```

Comment:
Wondering if we should use a different directory name, since this is the same one that master builds of docker use, and the format is changing in this PR?

Upgrades from docker master won't work correctly, right?